[Refactor/#78] 풋살장 위치 관련 리펙토링

Repaion24 opened the pull request:
## 📢 기능 설명 
<!-- 필요시 실행결과 스크린샷 첨부 -->
풋살장 Location 저장 방식을 (위도, 경도) 에서 MySQL에서 지원하는 공간 데이터 타입으로 리펙토링했습니다.

## 연결된 issue
<!-- 연결된 issue를 자동을 닫기 위해 아래 {이슈넘버}를 입력해주세요. -->
- close #78
<br>

## ✅ 체크리스트
- [x] PR 제목 규칙 잘 지켰는가? 
- [x] 추가/수정사항을 설명하였는가?
- [x] 이슈넘버를 적었는가?


📣 **To Reviewers**
---
<!-- 전달사항 -->
생각보다 코틀린-JPA 바닐라에서 공간 데이터 타입을 지원하지 않는 점들이 있어서 코드 수정하느라 좀 걸렸습니다...

Spring JPA에서 네이티브 쿼리보다는 JPQL이나 QueryDSL을 사용하라는 점도 있어,
QueryDSL은 공간 데이터 타입을 지원하지 않는 이유로 JPQL로 변환하여 작성해 보았습니다.
JPQL에서는 자동으로 페이징 처리를 안해주더라구요... 공부하는데 좀 걸렸습니다.

코드 중간중간 살피다 보니, 분명 사용해야 하는데 한번도 사용하지 않았다는 메서드들이 몇개 보이기에 관련해서 수정했습니다.

또한 distance 단위를 Km -> m로 변경했습니다. 
찾아보니, Km를 사용하여 ST_GeomFromText 로 변환할 때는, 지구 평균 반지름을 사용하여 degree로 변환하기에 오차가 좀 있다고 하더라구요! 그래서 변경했습니다.

만약 그 동안 DB에 풋살장 데이터를 저장해두신분들은... 아마 다 날아갈 예정입니다 😅
DB 칼럼이 변경되었기 때문에...

이 PR을 merge 한 후에는 ddl-auto: create 로 한번 날려야 정상적으로 동작할 것 같네요

혹시나 필요하신 분들은 아래와 같이 사용하시면 될 것 같습니다.

```json
{
    "name" :"스타디움01",
    "address":"서울시 용산구",
    "phoneNumber":"010-4444-4444",
    "description":"용산구 스타디움",
    "latitude":"37.53609444",
    "longitude":"126.9675222"
}

{
    "name": "스타디움02",
    "address": "강원도 춘천시 중앙로 100",
    "phoneNumber": "010-5555-5555",
    "description": "춘천시 중심에 위치한 대형 스타디움입니다.",
    "latitude": "37.8853",
    "longitude": "127.7295"
}

{
    "name": "스타디움03",
    "address": "대전광역시 유성구 문화로 200",
    "phoneNumber": "010-6666-6666",
    "description": "유성구에 위치한 최신 시설의 스타디움입니다.",
    "latitude": "36.3504",
    "longitude": "127.3845"
}

{
    "name": "스타디움04",
    "address": "부산광역시 해운대구 센텀동로 300",
    "phoneNumber": "010-7777-7777",
    "description": "부산 해운대구에 위치한 해변가 스타디움입니다.",
    "latitude": "35.1796",
    "longitude": "129.0756"
}

{
    "name": "스타디움05",
    "address": "전라남도 목포시 금남로 400",
    "phoneNumber": "010-8888-8888",
    "description": "목포시 중심부에 위치한 접근성이 좋은 스타디움입니다.",
    "latitude": "34.8051",
    "longitude": "126.3924"
}
```

Repaion24 committed:
refactor: #78 PointUtil 마이그레이션

Repaion24 committed:
refactor: #78 Stadium 도메인 수정

Repaion24 committed:
refactor: #78 Service, Request 리펙토링

Repaion24 committed:
refactor: #78 repository 리펙토링

Repaion24 committed:
refactor: #78 Service 리펙토링

Repaion24 committed:
refactor: #78 공간 함수 사용하도록 리펙토링

Repaion24 committed:
refactor: #78 인덱싱 추가

Repaion24 committed:
refactor: #78 softdelete 적용

Repaion24 committed:
fix: #78 마이그레이션 누락된 내용

NaMinhyeok commented on src/main/kotlin/com/grepp/nbe1_3_team04/stadium/domain/Stadium.kt:
```diff
@@ -88,11 +92,17 @@ class Stadium private constructor(
+        }
+    }
+
+    @PrePersist
+    @PreUpdate
+    fun ensureSrid() {
+        if (location.srid != 4326) {
+            location.srid = 4326
```
이거는 어떤 기능인가요 ??

Repaion24 replied:
코드 작성하다가 실행하고 알아낸 내용이었는데
```
Binary geometry function st_contains given two geometries of different srids: 4326 and 0, which should have been identical.
```
JPA로 생성된 location 칼럼이 SRID 값이 0을 가지더라구요. (일반적으로 SRID는 4326을 사용합니다.)

그래서 Entity가 영속화 되기 전에 SRID가 4326인지 확인하고, 그렇지 않은 경우 설정하는 코드입니다.

NaMinhyeok replied:
오.. 어렵네요 SRID가 뭔지부터 공부해봐야겠네요 아마도 Mysql의 Position 관련해서 있는거겠죠 ? 저도 한번 공부해보겠습니다!

Repaion24 replied:
SRID는 좌표체계입니다!  SRID 4326는 GPS 시스템에서도 많이 사용되는 표준 좌표 체계이구요.
좌표 체계가 동일하면, 동일한 모델 안에서 작업을 수행한다는 개념이라고 생각하시면 될 것 같아요.

지금 프로젝트에서는 SRID 4326을 사용했지만, 정확도가 좀 낮더라도 빠른 성능이나, 지도 상과 같은 2D 에 데이터를 표시할 때는 SRID 3857을 사용한다고 해요!

GIS좀 찾아보다가 알아낸 내용인데, 이런 것도 있구나 하시면 될 것 같습니다😊

NaMinhyeok replied:
위치 좌표 관련해서는 알고리즘도 중요하고 수학 중요하다고 하던데 뭐가 엄청 많네요.. 정보 알려주셔서 감사합니다 ! 공부하시느라 고생하셨습니다 !

NaMinhyeok commented on src/main/kotlin/com/grepp/nbe1_3_team04/stadium/repository/CustomStadiumRepositoryImpl.kt:
```diff
@@ -1,62 +1,42 @@
+@Repository
 class CustomStadiumRepositoryImpl(
-    private val queryFactory: JPAQueryFactory
+    @PersistenceContext private val entityManager: EntityManager
```
`@PersistenceContext` 가 추가된 이유가 있나요 ?

Repaion24 replied:
아.. 이거 맨 처음 작성할 때 SRID 관련 변경 사항 때문에 DB동기화가 필요해서 작성했던 건데,
지금은 데이터 조회만 해서 굳이 `@PersistenceContext`를 사용할 필요 없겠네요
`@Autowired` 사용하는게 더 어울리고 명확할 것 같습니다

NaMinhyeok commented on src/main/kotlin/com/grepp/nbe1_3_team04/stadium/service/StadiumServiceImpl.kt:
```diff
@@ -58,32 +59,40 @@ class StadiumServiceImpl(
         return stadiumRepository.findByAddressContainingIgnoreCase(address, pageable).map(StadiumsResponse::from)
     }
 
+    @Transactional
```
조회용 메서드라면 readOnly = true 쓰면 좋을 것 같아요 !

Repaion24 replied:
아 이거 해야지 해야지 하다가 까먹었네요 감사합니다

NaMinhyeok reviewed (approved):
새로운 기능이 생겼꾼여 고생하셨습니다!

MoonJaeGyeong commented on src/main/kotlin/com/grepp/nbe1_3_team04/stadium/domain/Stadium.kt:
```diff
@@ -46,8 +50,9 @@ class Stadium private constructor(
 
-    @Embedded
-    var position: Position = position
+    @field:NotNull
+    @Column(nullable = false, columnDefinition = "POINT")
```
제 기억상엔티티 단에서 저번에 민혁님이 얘기해주신 `@NotNull` 이 `@Column(nullable = false)` 를 대신해준다고 했던 거 같은데 두 번 들어가는 거면 하나는 지워도 좋을 것 같습니다!

Repaion24 replied:
예전 코드에서 긁어오다보니 그대로 넘어왔네요
수정하겠습니다!

MoonJaeGyeong commented on src/main/kotlin/com/grepp/nbe1_3_team04/stadium/util/PositionUtil.kt:
```diff
@@ -0,0 +1,14 @@
+
+object PositionUtil {
+
+    private val geometryFactory = GeometryFactory()
+
+    fun createPoint(latitude: Double, longitude: Double): Point {
+        return geometryFactory.createPoint(Coordinate(longitude, latitude))
```
위도와 경도를 받아서 point 라고 하는 객체를 만들어주는 라이브러리군요! 처음 봐요

MoonJaeGyeong reviewed (approved):
고생하셨습니다! 기능 추가를 코틀린에서 하시기 쉽지 않으셨을텐데 대단하시네요!

HMWG reviewed (approved):
고생 많으셨습니다!

이것저것 바뀐 부분이 많아서 고치는데 힘드셨을것같아요

뭔가 MySQL 함수를 사용해서 그런지 코드 자체는 간결해진 느낌이네요!

pyy2114 reviewed (approved):
고생하셨습니다!!

Repaion24 committed:
refactor: #78 코드 리뷰 반영

@field:NotNull, @Column(nullable = false) 중복 제거
영속성 체크 관련 제거
조회 매서드 readonly 반영
